kroxylicious/src/main/java/io/kroxylicious/proxy/service/ContributionManager.java

```diff
@@ -28,21 +29,26 @@ private ContributionManager() {
         this.loaderFunction = loaderFunction;
     }
 
-    public <T, C extends Context, S extends Contributor<T, C>> ConfigurationDefinition getDefinition(Class<S> contributorClass, String typeName) {
-        return findContributor(contributorClass, typeName, (typName, contributor) -> contributor.getConfigDefinition(typName));
+    public <S extends Contributor> ConfigurationDefinition getDefinition(Class<S> contributorClass,
+                                                                         String typeName) {
+        return (ConfigurationDefinition) findContributor(contributorClass, typeName, s -> new ConfigurationDefinition(s.getConfigType(), s.requiresConfiguration()));
     }
 
-    public <T, C extends Context, S extends Contributor<T, C>> T getInstance(Class<S> contributorClass, String typeName, C constructionContext) {
-        return findContributor(contributorClass, typeName, (typName, contributor) -> contributor.getInstance(typeName, constructionContext));
+    public <T, S extends Contributor> T getInstance(Class<S> contributorClass, String typeName,
+                                                    Context constructionContext) {
+        return (T) findContributor(contributorClass, typeName, contributor -> contributor.getInstance(constructionContext));
     }
 
-    private <T, C extends Context, S extends Contributor<T, C>, X> X findContributor(Class<S> contributorClass, String typeName, BiFunction<String, S, X> extractor) {
+    private <T, S extends Contributor<T, ?, ?>, X> X findContributor(Class<S> contributorClass, String typeName,
+                                                                     Function<S, X> extractor) {
         final Iterable<S> contributorsForClass = this.contributors.computeIfAbsent(contributorClass, loaderFunction);
         for (S contributor : contributorsForClass) {
-            if (contributor.contributes(typeName)) {
-                return extractor.apply(typeName, contributor);
+            if (contributor.getTypeName().equals(typeName)) {
```

Comment:
This is first-found-wins semantics? I don't think we want that. Better to throw. And this reveals that whole problem with type name... there's no mechanism for assigning those names (so conflicts are unavoidable), and there's no mechanism to further qualify the name to disambiguate a conflict. IIUC if two plugins claim the name `Foo` then you're stuck. This is why using fully qualified class names (with support for non-qualified class names when it's not ambiguous) seems to be like a better way to do it.

Reply:
Created #630 to make this change after we've finished with the API (should be applicable when we switch to using the classname)

Reply:
I'm strongly inclined to just do some validation during startup that each of the services loadable from each directory has a unique class name. While something more sophisticated might hold water conceptually I think it would be an unjustifiable time sink to tackle it right now. We could always relax things later on once we knew it was needed. I don't think this even this start-up time check is needed immediately, but we should open an issue or discussion for it.